File: Data Structures & Algorithms/three-integer-sum/submission-5.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = list()
        visited = set()
        nums.sort()

        for i in range(0, len(nums) - 2):
            l, r = i + 1, len(nums) - 1
            while l < r:
                if 0 == nums[i] + nums[l] + nums[r] and (nums[i], nums[l], nums[r]) not in visited:
                    temp = [nums[i], nums[l], nums[r]]
                    res.append(temp)
                    visited.add(tuple(temp))
                    l += 1
                    r -= 1
                elif 0 < nums[i] + nums[l] + nums[r]:
                    r -= 1
                else:
                    l += 1
        return res

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-5.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int zero = 0;
        int product = 1;
        for (int num : nums) {
            if (num == 0) {
                zero += 1;
                continue;
            }
            product *= num;
        }
        int[] res = new int[nums.length];
        if (zero > 1) {
            return res;
        }
        else if (zero == 1) {
            for (int i = 0; i < nums.length; i++) {
                if (nums[i] == 0) {
                    res[i] = product;
                    continue;
                }
                res[i] = 0;
            }
        }
        else {
            for (int i = 0; i < nums.length; i++) {
                res[i] = product / nums[i];
            }
        }
        return res;
    }
}  
